src/Microsoft.ML.Transforms/CustomMappingCatalog.cs

```diff
@@ -46,5 +46,64 @@ public static class CustomMappingCatalog
             where TSrc : class, new()
             where TDst : class, new()
             => new CustomMappingEstimator<TSrc, TDst>(catalog.GetEnvironment(), mapAction, contractName, inputSchemaDefinition, outputSchemaDefinition);
+
+        /// <summary>
+        /// Create a <see cref="StatefulCustomMappingEstimator{TSrc, TState, TDst}"/>, which applies a custom mapping of input columns to output columns,
+        /// while allowing a per-cursor state.
+        /// </summary>
+        /// <typeparam name="TSrc">The class defining which columns to take from the incoming data.</typeparam>
+        /// <typeparam name="TState">The type that describes per-cursor state.</typeparam>
+        /// <typeparam name="TDst">The class defining which new columns are added to the data.</typeparam>
+        /// <param name="catalog">The transform catalog.</param>
+        /// <param name="mapAction">The mapping action. In addition to the input and output objects, the action is given a state object that it can look at and/or modify.
+        /// If the resulting transformer needs to be save-able, the class defining <paramref name="mapAction"/> should implement
+        /// <see cref="StatefulCustomMappingFactory{TSrc, TState, TDst}"/> and needs to be decorated with
+        /// <see cref="CustomMappingFactoryAttributeAttribute"/> with the provided <paramref name="contractName"/>.
+        /// The assembly containing the class should be registered in the environment where it is loaded back
+        /// using <see cref="ComponentCatalog.RegisterAssembly(System.Reflection.Assembly, bool)"/>.</param>
+        /// <param name="stateInitAction">The action to initialize the state object, that is called once before the cursor is initialized.</param>
+        /// <param name="contractName">The contract name, used by ML.NET for loading the model.
+        /// If <see langword="null"/> is specified, resulting transformer would not be save-able.</param>
+        /// <param name="inputSchemaDefinition">Additional parameters for schema mapping between <typeparamref name="TSrc"/> and input data.
+        /// Useful when dealing with annotations.</param>
+        /// <param name="outputSchemaDefinition">Additional parameters for schema mapping between <typeparamref name="TDst"/> and output data.
+        /// Useful when dealing with annotations.</param>
+        public static StatefulCustomMappingEstimator<TSrc, TState, TDst> StatefulCustomMapping<TSrc, TState, TDst>(this TransformsCatalog catalog, Action<TSrc, TState, TDst> mapAction,
+            Action<TState> stateInitAction, string contractName, SchemaDefinition inputSchemaDefinition = null, SchemaDefinition outputSchemaDefinition = null)
```

Comment:
Can you please add Estimator characteristics (like the others) and add Exportable to Onnx: No?
(I am guessing this will never be exportable to onnx...is that correct?)

Reply:
Correct, it cannot be exportable to ONNX, since it is based on custom C# code specified by the user. The estimator documentation regarding this is in StatefulCustomMappingTransform.cs in line 332.

---
In reply to: [399067853](https://github.com/dotnet/machinelearning/pull/4676#discussion_r399067853) [](ancestors = 399067853)